pkg/cgroup/utils.go

```diff
@@ -1,45 +1,99 @@
 package cgroup
 
 import (
+	"errors"
+	"fmt"
+	"math"
 	"os"
+	"path"
 	"strconv"
 	"strings"
-
-	"k8s.io/klog/v2"
 )
 
-func readVariablesFromFile(filePath string) (map[string]uint64, error) {
-	data, err := os.ReadFile(filePath)
+type parseError struct {
+	Path string
+	File string
+	Err  error
+}
+
+func (e *parseError) Error() string {
+	return "unable to parse " + path.Join(e.Path, e.File) + ": " + e.Err.Error()
+}
+
+func (e *parseError) Unwrap() error { return e.Err }
+
+func parseKeyValue(t string) (string, uint64, error) {
+	parts := strings.SplitN(t, " ", 3)
+	if len(parts) != 2 {
+		return "", 0, fmt.Errorf("line %q is not in key value format", t)
+	}
+
+	value, err := parseUint(parts[1], 10, 64)
 	if err != nil {
-		return nil, err
-	}
-	res := map[string]uint64{}
-	for _, line := range strings.Split(string(data), "\n") {
-		parts := strings.Fields(line)
-		if len(parts) == 2 {
-			v, err := strconv.ParseUint(parts[1], 10, 64)
-			if err != nil {
-				klog.Warningf(`failed to parse cgroup stat line "%s": %s`, line, err)
-				continue
-			}
-			res[parts[0]] = v
+		return "", 0, err
+	}
+
+	return parts[0], value, nil
+}
+
+func parseUint(s string, base, bitSize int) (uint64, error) {
+	value, err := strconv.ParseUint(s, base, bitSize)
+	if err != nil {
+		intValue, intErr := strconv.ParseInt(s, base, bitSize)
+		// 1. Handle negative values greater than MinInt64 (and)
+		// 2. Handle negative values lesser than MinInt64
+		if intErr == nil && intValue < 0 {
+			return 0, nil
+		} else if errors.Is(intErr, strconv.ErrRange) && intValue < 0 {
+			return 0, nil
 		}
+
+		return value, err
 	}
-	return res, nil
+
+	return value, nil
 }
 
-func readIntFromFile(filePath string) (int64, error) {
-	data, err := os.ReadFile(filePath)
+func openFile(dirPath, fileName string) (*os.File, error) {
```

Comment:
Why not simply use `os.Open`?

Reply:
openFile is a small helper specific for cgroup files, adds permission in one place.